openedx/core/djangoapps/content_libraries/api.py

```diff
@@ -410,35 +503,69 @@ def delete_library(library_key):
     # system, which is a better state than having a reference to a library with
     # no backing blockstore bundle.
     ref.delete()
+    CONTENT_LIBRARY_DELETED.send(sender=None, library_key=ref.library_key)
     try:
         delete_bundle(bundle_uuid)
     except:
         log.exception("Failed to delete blockstore bundle %s when deleting library. Delete it manually.", bundle_uuid)
         raise
 
 
-def get_library_blocks(library_key):
+def get_library_blocks(library_key, text_search=None):
     """
     Get the list of top-level XBlocks in the specified library.
 
     Returns a list of LibraryXBlockMetadata objects
     """
-    ref = ContentLibrary.objects.get_by_key(library_key)
-    lib_bundle = LibraryBundle(library_key, ref.bundle_uuid, draft_name=DRAFT_NAME)
-    usages = lib_bundle.get_top_level_usages()
-    blocks = []
-    for usage_key in usages:
-        # For top-level definitions, we can go from definition key to usage key using the following, but this would not
-        # work for non-top-level blocks as they may have multiple usages. Top level blocks are guaranteed to have only
-        # a single usage in the library, which is part of the definition of top level block.
-        def_key = lib_bundle.definition_for_usage(usage_key)
-        blocks.append(LibraryXBlockMetadata(
-            usage_key=usage_key,
-            def_key=def_key,
-            display_name=get_block_display_name(def_key),
-            has_unpublished_changes=lib_bundle.does_definition_have_unpublished_changes(def_key),
-        ))
-    return blocks
+    metadata = None
+    if LibraryBlockIndexer.indexing_is_enabled():
+        try:
+            filter_terms = {
+                'library_key': [str(library_key)],
+                'is_child': [False],
+            }
+            metadata = [
+                {
+                    **item,
+                    "id": LibraryUsageLocatorV2.from_string(item['id']),
+                }
+                for item in LibraryBlockIndexer.get_items(filter_terms=filter_terms, text_search=text_search)
+                if item is not None
+            ]
+        except (ConnectionError) as e:
+            log.exception(e)
+
+    if metadata is None:
```

Comment:
What if an indexed search returns no hits?  It looks like we'll be returning all items (and from the slow API), which is probably not the desired behavior.  What is the downside to using an "else", here?  Are you protecting for unindexed libraries?

Reply:
If the indexed search returns no hits, `metadata` would be an empty list `[]`, and skip this block.
Using this check instead of `else` so that this block gets executed if indexing is enabled, but connection to elasticsearch fails (cf. the no-op on `ConnectionError` in the line above)

Reply:
Ah, of course!  Makes total sense. +1